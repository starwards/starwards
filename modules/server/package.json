{
    "name": "@starwards/server",
    "version": "0.0.0",
    "private": true,
    "license": "SEE LICENSE IN LICENSE.md",
    "files": [
        "cjs"
    ],
    "scripts": {
        "clean": "rimraf ./cjs",
        "postbuild": "npm pack --quiet",
        "prebuild": "npm run clean",
        "build": "tsc -p ./tsconfig.runtime.json",
        "start": "node -r ts-node/register/transpile-only ./src/dev.ts",
        "loadtest": "colyseus-loadtest loadtest/bot.ts --room space --endpoint ws://localhost:8080",
        "test": "jest"
    },
    "dependencies": {
        "@colyseus/monitor": "^0.14.22",
        "@colyseus/ws-transport": "^0.14.21",
        "async-mqtt": "^2.6.3",
        "colyseus": "^0.14.23",
        "express": "^4.18.1",
        "express-async-handler": "^1.2.0",
        "express-basic-auth": "^1.2.1"
    }
}
