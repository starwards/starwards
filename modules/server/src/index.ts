export {State} from './rooms/State';
export {Entity} from './rooms/Entity';
