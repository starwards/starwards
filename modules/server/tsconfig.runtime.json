{
    "extends": "./tsconfig.json",
    "compilerOptions": {
        "module": "commonjs",
        "outDir": "cjs",
        "experimentalDecorators": true,
        "paths": { "@starwards/core": ["../core/cjs"] }
    },
    "exclude": ["node_modules", "src/**/*.spec.ts"]
}
