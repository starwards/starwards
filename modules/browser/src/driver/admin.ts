import { GameRoom, adminProperties, cmdSender } from '@starwards/model/src';

export function AdminDriver(adminRoom: GameRoom<'admin'>) {
    const shouldGameBeRunning = cmdSender(adminRoom, adminProperties.shouldGameBeRunning);
    return {
        state: adminRoom.state,
        stopGame: () => shouldGameBeRunning(false),
        startGame: () => shouldGameBeRunning(true),
    };
}
