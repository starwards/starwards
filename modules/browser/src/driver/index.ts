import { AdminDriver } from './admin';
import { Client } from 'colyseus.js';
import { ShipDriver } from './ship';
import { SpaceDriver } from './space';
import { schemaClasses } from '@starwards/model';
import { waitForEvents } from './async-utils';

export { DriverNumericApi, NumberMapDriver } from './utils';

export type AdminDriver = ReturnType<typeof AdminDriver>;
export type ShipDriver = ReturnType<typeof ShipDriver>;
export type SpaceDriver = ReturnType<typeof SpaceDriver>;

// const ENDPOINT = 'ws:' + window.location.href.substring(window.location.protocol.length);
const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
const ENDPOINT = protocol + '//' + window.location.host; // + '/';

export class Driver {
    private adminDriver: Promise<AdminDriver> | null = null;
    private spaceDriver: Promise<SpaceDriver> | null = null;
    private shipDrivers = new Map<string, Promise<ShipDriver>>();

    private client = new Client(ENDPOINT);

    async isActiveGame() {
        const rooms = await this.client.getAvailableRooms('space');
        return !!rooms.length;
    }

    /**
     * Returns a finite iterator for the IDs of the ships currently active
     */
    async getCurrentShipIds(): Promise<Iterable<string>> {
        const rooms = await this.client.getAvailableRooms('ship');
        return rooms.map((r) => r.roomId);
    }

    /**
     * infinite iterator
     */
    async *getUniqueShipIds() {
        const ships = new Set<string>();
        while (true) {
            for (const shipId of await this.getCurrentShipIds()) {
                if (!ships.has(shipId)) {
                    ships.add(shipId);
                    yield shipId;
                }
            }
            await new Promise((res) => setTimeout(res, 500));
        }
    }

    /**
     * constantly scan for new ships and return when current ship is found
     * @param shipToWaitFor id of the ship to wait for
     */
    async waitForShip(shipToWaitFor: string): Promise<void> {
        for await (const shipId of this.getUniqueShipIds()) {
            if (shipToWaitFor === shipId) {
                return;
            }
        }
    }

    async getShipDriver(shipId: string): Promise<ShipDriver> {
        if (this.shipDrivers.has(shipId)) {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            return this.shipDrivers.get(shipId)!;
        }
        const shipDriver = this.makeShipDriver(shipId);
        this.shipDrivers.set(shipId, shipDriver);
        return shipDriver;
    }

    private async makeShipDriver(shipId: string) {
        const room = await this.client.joinById(shipId, {}, schemaClasses.ship);
        const pendingEvents = [];
        if (!room.state.chainGun) {
            pendingEvents.push('chainGun');
        }
        if (!room.state.constants) {
            pendingEvents.push('constants');
        }
        await waitForEvents(room.state.events, pendingEvents);
        return ShipDriver(room);
    }

    async getSpaceDriver(): Promise<SpaceDriver> {
        if (this.spaceDriver) {
            return await this.spaceDriver;
        }
        this.spaceDriver = this.client.join('space', {}, schemaClasses.space).then(SpaceDriver);
        return await this.spaceDriver;
    }

    async getAdminDriver(): Promise<AdminDriver> {
        if (this.adminDriver) {
            return await this.adminDriver;
        }
        this.adminDriver = this.client.join('admin', {}, schemaClasses.admin).then(AdminDriver);
        return await this.adminDriver;
    }
}
