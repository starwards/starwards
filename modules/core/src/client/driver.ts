import { AdminState, EventEmitter, schemaClasses } from '..';
import { Client, Room } from 'colyseus.js';
import { ConnectionManager, ConnectionStateEvent } from './connection-manager';

import { AdminDriver } from './admin';
import { SchemaConstructor } from 'colyseus.js/lib/serializer/SchemaSerializer';
import { ShipDriver } from './ship';
import { SpaceDriver } from './space';
import { waitForEvents } from '../async-utils';

export type ShipDriverRead = Pick<ShipDriver, 'state' | 'events'>;

export type AdminDriver = ReturnType<ReturnType<typeof AdminDriver>>;

const joinRetries = 3;

export function getColyseusEndpoint(location: { protocol: string; host: string }) {
    return (location.protocol === 'https:' ? 'wss:' : 'ws:') + '//' + location.host + '/colyseus';
}
export class Driver {
    private connectionManager = new ConnectionManager(() => {
        this.adminDriver = this.joinRoom('admin', schemaClasses.admin)
            .then(this.hookAdminRoomLifecycle)
            .then(AdminDriver(this.httpEndpoint));
        return this.adminDriver;
    });
    public get connectionStatus(): EventEmitter<{ [k in ConnectionStateEvent]: void }> {
        return this.connectionManager.events;
    }
    public get isConnected() {
        return this.connectionManager.stateConnected;
    }
    public get errorMessage(): string | null {
        return this.connectionManager.getErrorMessage();
    }
    public httpEndpoint: string;
    private rooms: Client;

    private adminDriver: Promise<AdminDriver> | null = null;
    private spaceDriver: Promise<SpaceDriver> | null = null;
    private shipDrivers = new Map<string, Promise<ShipDriver>>();

    private hookRoomLifecycle = <R extends Room>(room: R) => {
        room.onError((code, message) => this.connectionManager.onConnectionError({ code, message }));
        if (this.connectionManager.isDestroyed) {
            void room.leave(true);
            throw new Error(`client destroyed while room joined`);
        }
        this.connectionManager.events.once('exit:connected', () => void room.leave(true));
        this.connectionManager.events.once('destroyed', () => void room.leave(true));
        return room;
    };

    private hookAdminRoomLifecycle = (room: Room<AdminState>) => {
        room.onLeave(() => {
            if (!this.connectionManager.isDestroyed) {
                this.connectionManager.onConnectionError(new Error(`disconnected from server`));
            }
        });
        return this.hookRoomLifecycle(room);
    };

    /**
     * @param location window.location compatible object
     */
    constructor(location: { protocol: string; host: string }) {
        const colyseusEndpoint = getColyseusEndpoint(location);
        this.httpEndpoint = location.protocol + '//' + location.host;
        this.rooms = new Client(colyseusEndpoint);
        this.connectionManager.events.on('exit:connected', this.clearCache);
        void this.onGameStateChange(async () => {
            if (!(await this.isActiveGame())) {
                this.spaceDriver = null;
                this.shipDrivers.clear();
            }
        });
    }

    private async joinRoom<T>(roomId: string, rootSchema: SchemaConstructor<T>): Promise<Room<T>> {
        for (let i = 1; i < joinRetries; i++) {
            try {
                return await this.rooms.joinById(roomId, {}, rootSchema);
                // eslint-disable-next-line no-empty
            } catch (e) {}
        }
        return await this.rooms.joinById(roomId, {}, rootSchema);
    }

    clearCache = () => {
        this.adminDriver = null;
        this.spaceDriver = null;
        this.shipDrivers.clear();
    };

    connect() {
        this.connectionManager.connect();
        return this;
    }
    destroy() {
        this.connectionManager.destroy();
    }

    onGameStateChange(cb: () => unknown): () => void {
        const wrappedListener: () => unknown = async () => {
            try {
                if (!this.connectionManager.isDestroyed) {
                    await cb();
                }
            } catch (e) {
                void 0;
            }
        };
        let abort = false;
        let unRegister = () => undefined as unknown;
        this.connectionManager.events.on('*', wrappedListener);
        void (async () => {
            while (!abort && !this.connectionManager.isDestroyed) {
                try {
                    await this.connectionManager.waitForConnected();
                } catch (e) {
                    continue;
                }
                if (!this.adminDriver) {
                    throw new Error('connected, but missing adminDriver');
                }
                const driver = await this.adminDriver;
                await wrappedListener();
                driver.events.on('**', wrappedListener);
                unRegister = () => driver.events.off('**', wrappedListener);
                await waitForEvents(this.connectionManager.events, ['exit:connected']);
                unRegister();
            }
        })();
        return () => {
            abort = true;
            this.connectionManager.events.off('*', wrappedListener);
            unRegister();
        };
    }
    async isActiveGame() {
        await this.connectionManager.waitForConnected();
        if (!this.adminDriver) {
            throw new Error('connected, but missing adminDriver');
        }
        return (await this.adminDriver).state.isGameRunning;
    }

    /**
     * Returns a finite iterator for the IDs of the ships currently active
     */
    async getCurrentShipIds(): Promise<Iterable<string>> {
        await this.connectionManager.waitForConnected();
        if (!this.adminDriver) {
            throw new Error('connected, but missing adminDriver');
        }
        return (await this.adminDriver).state.shipIds;
    }

    async doesShipExist(shipId: string): Promise<boolean> {
        await this.connectionManager.waitForConnected();
        if (!this.adminDriver) {
            throw new Error('connected, but missing adminDriver');
        }
        return (await this.adminDriver).state.shipIds.includes(shipId);
    }

    /**
     * infinite iterator
     */
    async *getUniqueShipIds() {
        const ships = new Set<string>();
        while (!this.connectionManager.isDestroyed) {
            for (const shipId of await this.getCurrentShipIds()) {
                if (!ships.has(shipId)) {
                    ships.add(shipId);
                    yield shipId;
                }
            }
            await new Promise((res) => setTimeout(res, 500));
        }
    }

    /**
     * constantly scan for a game and return when a game is running
     */
    async waitForGame(): Promise<void> {
        while (!(await this.isActiveGame())) {
            await new Promise((res) => setTimeout(res, 500));
        }
    }
    /**
     * constantly scan for new ships and return when current ship is found
     * @param shipToWaitFor id of the ship to wait for
     */
    async waitForShip(shipToWaitFor: string): Promise<void> {
        for await (const shipId of this.getUniqueShipIds()) {
            if (shipToWaitFor === shipId) {
                return;
            }
        }
        throw new Error('client destroyed before ship appeared');
    }

    async getShipDriver(shipId: string): Promise<ShipDriver> {
        await this.connectionManager.waitForConnected();
        if (this.shipDrivers.has(shipId)) {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            return this.shipDrivers.get(shipId)!;
        }
        const shipDriver = this.makeShipDriver(shipId);
        this.shipDrivers.set(shipId, shipDriver);
        return shipDriver;
    }

    private async makeShipDriver(shipId: string) {
        try {
            await this.waitForShip(shipId);
            const room = await this.joinRoom(shipId, schemaClasses.ship).then(this.hookRoomLifecycle);
            return await ShipDriver(room);
        } catch (e) {
            const error = new Error('failed making ship driver', { cause: e });
            this.connectionManager.onConnectionError(error);
            throw error;
        }
    }

    async getSpaceDriver(): Promise<SpaceDriver> {
        await this.connectionManager.waitForConnected();
        if (this.spaceDriver) {
            return await this.spaceDriver;
        }
        try {
            this.spaceDriver = this.joinRoom('space', schemaClasses.space)
                .then(this.hookRoomLifecycle)
                .then(SpaceDriver);
            return await this.spaceDriver;
        } catch (e) {
            const error = new Error('failed making space driver', { cause: e });
            this.connectionManager.onConnectionError(error);
            throw error;
        }
    }

    async getAdminDriver(): Promise<AdminDriver> {
        await this.connectionManager.waitForConnected();
        if (!this.adminDriver) {
            throw new Error('connected, but missing adminDriver');
        }
        return this.adminDriver;
    }
}
