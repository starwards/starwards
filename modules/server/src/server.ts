// tslint:disable no-console
/*
 * embed webpack-dev-server
 */

import { Server, matchMaker } from 'colyseus';
import * as http from 'http';
import express = require('express');
import basicAuth = require('express-basic-auth');
import { monitor } from '@colyseus/monitor';
import { SpaceRoom } from './space/room';
import { NextHandleFunction } from 'connect';
import { AdminRoom } from './admin/room';
import { ShipRoom } from './ship/room';
import { GameManager } from './admin/game-manager';

process.on('uncaughtException', function (err) {
    console.error(new Date().toUTCString() + ' uncaughtException:', err.message);
    console.error(err.stack);
    // process.exit(1);
});

export async function server(port: number, staticDir: string, handlers?: NextHandleFunction[]) {
    const app = express();
    app.use(express.json());
    const gameServer = new Server({ server: http.createServer(app) });

    gameServer
        .define('space', SpaceRoom)
        .on('create', (room) => console.log('SpaceRoom created:', room.roomId))
        .on('dispose', (room) => console.log('SpaceRoom disposed:', room.roomId))
        .on('join', (room, client) => console.log('SpaceRoom:', client.sessionId, 'joined', room.roomId))
        .on('leave', (room, client) => console.log('SpaceRoom:', client.sessionId, 'left', room.roomId));

    gameServer
        .define('admin', AdminRoom)
        .on('create', (room) => console.log('AdminRoom created:', room.roomId))
        .on('dispose', (room) => console.log('AdminRoom disposed:', room.roomId))
        .on('join', (room, client) => console.log('AdminRoom:', client.sessionId, 'joined', room.roomId))
        .on('leave', (room, client) => console.log('AdminRoom:', client.sessionId, 'left', room.roomId));

    gameServer
        .define('ship', ShipRoom)
        .on('create', (room) => console.log('ship created:', room.roomId))
        .on('dispose', (room) => console.log('ship disposed:', room.roomId))
        .on('join', (room, client) => console.log('ship:', client.sessionId, 'joined', room.roomId))
        .on('leave', (room, client) => console.log('ship:', client.sessionId, 'left', room.roomId))
        .enableRealtimeListing();

    if (handlers) {
        app.use(...handlers);
    }

    app.use('/', express.static(staticDir));

    // add colyseus monitor
    const auth = basicAuth({ users: { admin: 'admin' }, challenge: true });
    app.use('/colyseus', auth, monitor());

    await gameServer.listen(port);
    // tslint:disable-next-line:no-console
    console.log(`Listening on port ${port}`);

    const gameManager = new GameManager();
    await matchMaker.createRoom('admin', { manager: gameManager }); // create a room
}
