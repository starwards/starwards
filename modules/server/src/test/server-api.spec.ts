import { HTTP_CONFLICT_STATUS } from '../server';
import { makeDriver } from './driver';
import supertest from 'supertest';

describe('server-API', () => {
    const gameDriver = makeDriver();

    test('start and stop a game (using gameManager API directly)', async () => {
        expect(gameDriver.gameManager.state.isGameRunning).toBe(false);
        await supertest(gameDriver.httpServer).post('/start-game').send({ mapName: 'two_vs_one' }).expect(200);
        expect(gameDriver.gameManager.state.isGameRunning).toBe(true);
        await supertest(gameDriver.httpServer).post('/stop-game').expect(200);
        expect(gameDriver.gameManager.state.isGameRunning).toBe(false);
    });

    test('save game when not running returns 409 conflict', async () => {
        const response = await supertest(gameDriver.httpServer).post('/save-game');
        expect(response.status).toEqual(HTTP_CONFLICT_STATUS);
    });

    test('save game returns same string', async () => {
        gameDriver.pauseGameCommand();
        await supertest(gameDriver.httpServer).post('/start-game').send({ mapName: 'test_map_1' }).expect(200);
        const response = await supertest(gameDriver.httpServer).post('/save-game');
        expect(response.status).toEqual(200);
        expect(response.text).toMatchSnapshot('test_map_1-save-game');
    });
});
