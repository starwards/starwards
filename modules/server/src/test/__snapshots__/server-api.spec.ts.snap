// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server-API save game functionality regression test: compatibility with saved games from older version: test_map_1-save-game 1`] = `"gKp0ZXN0X21hcF8xgQH/AYACgQP/AoAApEdWVFME/wOABYEGggeDCP8EgQCCAIOkR1ZUU4QJhQAASEKGCor/iwAAAACHAAAAAIgAAAAAgKlTcGFjZXNoaXCTAAAAAJQAAAAAlQAAAACWAAAAAJcAAAAAmAAAAACZAJsQDgAAjQuODI8NkA6RD5IQmhH/Bf8G/weAAKZhc3RybzESgAGmYXN0cm8yE4ACpmFzdHJvMxT/CIAApEdWVFMV/wmAAAAAAIEAAAAA/wqAAAAAAIEAAAAA/wuAFv8MgAAXgAEYgAIZgAMagAQbgAUc/w2AAAAAAIEAggAAAACDAAAAAIQAAAAAhgAAAACHAYgd/w6BAAAAAIAe/w+BzegDggCDAACAP4Af/xCDAAAAAIQAAAAAhSCGAAAAAIAhgQGCAf8RgCKBI/8SgQCCAIOmYXN0cm8xhCSFAABIQ4YlhwAAAACIAAAAAICoQXN0ZXJvaWSJZAD/E4EAggCDpmFzdHJvMoQmhQAAr0OGJ4cAAAAAiAAAAACAqEFzdGVyb2lkiWQA/xSBAIIAg6Zhc3RybzOEKIUAAEhChimHAAAAAIgAAAAAgKhBc3Rlcm9pZIlkAP8VgQCCAIOkR1ZUU4QqhQAASEKGK4oAiwAAAACHAAAAAIgAAAAAgKlTcGFjZXNoaXCMrmRyYWdvbmZseS1TRjIy/xaAALByb3RhdGlvbkNhcGFjaXR5GYABsnJvdGF0aW9uRW5lcmd5Q29zdMvsUbgeheuxP4ACr21heENoYWluR3VuQW1tb80QDv8XgAAAAACBAAAAAIIAALTCgwAAAACEAACAP4Us/xiAAAAAAIEAAAAAggAAtEKDAAAAAIQAAIA/hS3/GYAAAAAAgQAAAACCAAAAAIMAAAAAhAAAgD+FLv8agAAAAACBAAAAAIIAAAAAgwAAAACEAACAP4Uv/xuAAAAAAIEAAAAAggAANEODAAAAAIQAAIA/hTD/HIAAAAAAgQAAAACCAAA0Q4MAAAAAhAAAgD+FMf8dgDL/HoAz/x+ANP8ggAAAAACBAAAAAP8hgDX/IoAANoABN4ACOIADOYAEOoAFO4AGPIAHPYAIPoAJP4AKQIALQf8jgEL/JIAAAPpEgQAA+kT/JYAAAAAAgQAAAAD/JoAAAHpEgQAAekT/J4AAAAAAgQAAAAD/KIAAgDtFgQAA+sT/KYAAAAAAgQAAAAD/KoAAAAAAgQAAAAD/K4AAAAAAgQAAAAD/LIBD/y2ARP8ugEX/L4BG/zCAR/8xgEj/MoAAsGJ1bGxldHNQZXJTZWNvbmQUgAGrYnVsbGV0U3BlZWTN6AOAArZidWxsZXREZWdyZWVzRGV2aWF0aW9uAYADrW1heFNoZWxsUmFuZ2XNiBOABK1taW5TaGVsbFJhbmdlzegDgAWtc2hlbGxSYW5nZUFpbc3oA4AGr2V4cGxvc2lvblJhZGl1cwqAB7dleHBsb3Npb25FeHBhbnNpb25TcGVlZCiACLVleHBsb3Npb25EYW1hZ2VGYWN0b3IUgAm0ZXhwbG9zaW9uQmxhc3RGYWN0b3IBgAqoZGFtYWdlNTAUgAu+Y29tcGxldGVEZXN0cnVjdGlvblByb2JhYmlsaXR5y5qZmZmZmbk//zOAAKhkYW1hZ2U1MBSAAapiYXNpY1JhbmdlzbgLgAKvcmFuZ2VFYXNlRmFjdG9yy5qZmZmZmck/gAOwbWFsZnVuY3Rpb25SYW5nZc3cBf80gACvZW5lcmd5UGVyU2Vjb25kBYABqW1heEVuZXJnec3oA4ACsm1heEFmdGVyQnVybmVyRnVlbM2IE4ADsWFmdGVyQnVybmVyQ2hhcmdlFIAEtWFmdGVyQnVybmVyRW5lcmd5Q29zdMvsUbgeheuxP4AFqGRhbWFnZTUwFP81gACybWF4VGFyZ2V0QWltT2Zmc2V0HoABrmFpbU9mZnNldFNwZWVkD4ACrG1heFR1cm5TcGVlZFqAA7VvZmZzZXRCcm9rZW5UaHJlc2hvbGTLMzMzMzMz4z+ABKhkYW1hZ2U1MFqABahtYXhTcGVlZM0sAYAGt21heFNwZWVkRnJvbUFmdGVyQnVybmVyzSwB/zaAAIC7RP83gACAu0T/OIAAgLtE/zmAAIC7RP86gACAu0T/O4AAgLtE/zyAAIC7RP89gACAu0T/PoAAgLtE/z+AAIC7RP9AgACAu0T/QYAAgLtE/0KAAK5udW1iZXJPZlBsYXRlcwyAAahoZWFsUmF0ZQOAAq5wbGF0ZU1heEhlYWx0aM3cBf9DgACtbWF4QW5nbGVFcnJvci2AAahjYXBhY2l0eTKAAqplbmVyZ3lDb3N0y+xRuB6F67E/gAOrc3BlZWRGYWN0b3IDgASzYWZ0ZXJCdXJuZXJDYXBhY2l0ec0sAYAFt2FmdGVyQnVybmVyRWZmZWN0RmFjdG9yAYAGqGRhbWFnZTUwD4AHvmNvbXBsZXRlRGVzdHJ1Y3Rpb25Qcm9iYWJpbGl0ecuamZmZmZm5P/9EgACtbWF4QW5nbGVFcnJvci2AAahjYXBhY2l0eTKAAqplbmVyZ3lDb3N0y+xRuB6F67E/gAOrc3BlZWRGYWN0b3IDgASzYWZ0ZXJCdXJuZXJDYXBhY2l0ec0sAYAFt2FmdGVyQnVybmVyRWZmZWN0RmFjdG9yAYAGqGRhbWFnZTUwD4AHvmNvbXBsZXRlRGVzdHJ1Y3Rpb25Qcm9iYWJpbGl0ecuamZmZmZm5P/9FgACtbWF4QW5nbGVFcnJvci2AAahjYXBhY2l0eTKAAqplbmVyZ3lDb3N0y+xRuB6F67E/gAOrc3BlZWRGYWN0b3IDgASzYWZ0ZXJCdXJuZXJDYXBhY2l0ec0sAYAFt2FmdGVyQnVybmVyRWZmZWN0RmFjdG9yAYAGqGRhbWFnZTUwD4AHvmNvbXBsZXRlRGVzdHJ1Y3Rpb25Qcm9iYWJpbGl0ecuamZmZmZm5P/9GgACtbWF4QW5nbGVFcnJvci2AAahjYXBhY2l0eTKAAqplbmVyZ3lDb3N0y+xRuB6F67E/gAOrc3BlZWRGYWN0b3IDgASzYWZ0ZXJCdXJuZXJDYXBhY2l0ec0sAYAFt2FmdGVyQnVybmVyRWZmZWN0RmFjdG9yAYAGqGRhbWFnZTUwD4AHvmNvbXBsZXRlRGVzdHJ1Y3Rpb25Qcm9iYWJpbGl0ecuamZmZmZm5P/9HgACtbWF4QW5nbGVFcnJvci2AAahjYXBhY2l0eTKAAqplbmVyZ3lDb3N0y+xRuB6F67E/gAOrc3BlZWRGYWN0b3IDgASzYWZ0ZXJCdXJuZXJDYXBhY2l0ec0sAYAFt2FmdGVyQnVybmVyRWZmZWN0RmFjdG9yAYAGqGRhbWFnZTUwD4AHvmNvbXBsZXRlRGVzdHJ1Y3Rpb25Qcm9iYWJpbGl0ecuamZmZmZm5P/9IgACtbWF4QW5nbGVFcnJvci2AAahjYXBhY2l0eTKAAqplbmVyZ3lDb3N0y+xRuB6F67E/gAOrc3BlZWRGYWN0b3IDgASzYWZ0ZXJCdXJuZXJDYXBhY2l0ec0sAYAFt2FmdGVyQnVybmVyRWZmZWN0RmFjdG9yAYAGqGRhbWFnZTUwD4AHvmNvbXBsZXRlRGVzdHJ1Y3Rpb25Qcm9iYWJpbGl0ecuamZmZmZm5Pw=="`;
